### PR TITLE
Cache proposals

### LINKED ISSUE #1517: Sequential Proposal follow up -Caching
This is follow up work from the PR here --> https://github.com/decentdao/fractal-interface/pull/1506 

As discussed with Adam at standup, there's room to extend proposals loading sequentially to add

- Cache completed proposals in local storage, so we don't need to re-fetch them.

My only additional thought is if completed proposals load faster, couldn't we get proposals coming in in the wrong order or confusingly.

Excited to pick this up though

### LINKED ISSUE #1567: Show more accurate loading indicators when Proposals are sequentially loading in
When proposals are loading in, we should improve the experience in two ways:

1. Don't show "No Proposals" component when the first proposal is loading in.
2. Indicate to the user that proposals are still coming in, until they're finished.

Probably need some design assistance for the new component for 2. above.

### PR DESCRIPTION
**Pull request by DarksightKellar**
Closes #1517 

This PR adds functionality for the app to cache "fossilised" proposals:
```
const isProposalFossilized =
          proposal.state === FractalProposalState.CLOSED ||
          proposal.state === FractalProposalState.EXECUTED ||
          proposal.state === FractalProposalState.FAILED ||
          proposal.state === FractalProposalState.EXPIRED ||
          proposal.state === FractalProposalState.REJECTED;
```

These proposals' data are basically immutable and can be safely cached eternally. This speeds up general proposal loading significantly as only "active" proposals need to be sequentially loaded.

I have also made it so that the loading indicator continues to show at the bottom of the proposal list if sequential loading isn't done but some proposals have already been loaded (and thus potentially closes #1567):

Screenshot 2024-05-30 at 15 36 37

Screenshot 2024-05-30 at 15 36 45

Also, the "No Activity" card will only show if there is truly no activity.

### CODE REVIEW

**Comment by adamgall** [comment]
@DarksightKellar should #1567 be linked to this PR as well?

**Comment by DarksightKellar** [comment]
> @DarksightKellar should #1567 be linked to this PR as well?

@adamgall  I just edited the description to include that

**Comment by DarksightKellar** [comment]
Tagging and requesting review from @decentdao/design to validate sequential loading indicator UX

**Review by xanaramoss — APPROVED** [review]
Looks and feels good to me. 🤙

File: src/hooks/utils/cache/useLocalStorage.ts [hunk 1]
```diff
@@ -1,7 +1,26 @@
...
+function proposalObjectReviver(key: any, value: any) {
+  if (typeof value === 'string') {
+    if (value.startsWith('bigint:')) return BigInt(value.substring(7));
+    const isoStringRegex = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d{3}Z$/;
+    if (isoStringRegex.test(value)) return parseISO(value);
+  }
+
+  return value;
+}
+
```
Comment (Da-Colon) [inline src/hooks/utils/cache/useLocalStorage.ts:23]: o interesting, didn't know about reviver functions this is cool. Doesn't effect my thoughts of updates, if anything, enhances it. nice

**Review by Da-Colon — APPROVED** [review]

File: src/hooks/utils/cache/useLocalStorage.ts [hunk 2]
```diff
@@ -1,7 +1,26 @@
...
+  return typeof value === 'bigint'
+    ? `bigint:${value.toString()}`
+    : value instanceof Date
+      ? value.toISOString()
+      : value;
+}
+
+function proposalObjectReviver(key: any, value: any) {
+  if (typeof value === 'string') {
+    if (value.startsWith('bigint:')) return BigInt(value.substring(7));
```
Comment (adamgall) [inline src/hooks/utils/cache/useLocalStorage.ts:16]: What is the `7` magic number here?

**Review by adamgall — DISMISSED** [review]
@DarksightKellar , at first glance this code looks good!

It always requires caution though, when deploying "caching" software. For example, what if after we've deployed this, we realize there's a bug? Will we be able to easily bust the cache?

I recommend adding some kind of "versioning" in the cache logic, so that we can bump a hardcoded number in the codebase to force all clients to recognize that "the data in my cache is stale, I should trash it all start fresh".

**Comment by Da-Colon** [comment]
> @DarksightKellar , at first glance this code looks good!
> 
> It always requires caution though, when deploying "caching" software. For example, what if after we've deployed this, we realize there's a bug? Will we be able to easily bust the cache?
> 
> I recommend adding some kind of "versioning" in the cache logic, so that we can bump a hardcoded number in the codebase to force all clients to recognize that "the data in my cache is stale, I should trash it all start fresh".

Yeah this versioning was some thing I was going to add to the local storage update. But could be good to add this here prior to myupdate.

 Gonna revisit my code for it shortly (after this is merged in) as now I want to play with those reviver function to clean up stuff even more (I hope).

**Comment by DarksightKellar** [comment]
> @DarksightKellar , at first glance this code looks good!
> 
> It always requires caution though, when deploying "caching" software. For example, what if after we've deployed this, we realize there's a bug? Will we be able to easily bust the cache?
> 
> I recommend adding some kind of "versioning" in the cache logic, so that we can bump a hardcoded number in the codebase to force all clients to recognize that "the data in my cache is stale, I should trash it all start fresh".

Yeah, very good point.

File: src/hooks/utils/cache/useLocalStorage.ts (hunk 2 above)
Comment (DarksightKellar) [reply src/hooks/utils/cache/useLocalStorage.ts:16]: Length of `bigint:` prefix

File: src/hooks/utils/cache/cacheDefaults.ts [hunk 3]
```diff
@@ -47,6 +89,18 @@ export const CACHE_DEFAULTS: IndexedObject = {
  * mechanism should internally combine the key set in CacheKeys
  * along with the chainId.
  */
-export function keyInternal(chainId: number, key: string): string {
-  return 'fract_' + chainId + '_' + key;
+export function keyInternal(chainId: number, key: string, isVersioned: boolean): string {
+  return `${isVersioned ? VERSION_PREFIX : ''}fract_${chainId}_${key}`;
+}
+
+// @dev where's a good place to call this?
```
Comment (DarksightKellar) [inline src/hooks/utils/cache/cacheDefaults.ts:96]: @decentdao/engineering halp pls

Comment (adamgall) [reply src/hooks/utils/cache/cacheDefaults.ts:96]: @Da-Colon can u halp w/ this plz?

Comment (DarksightKellar) [reply src/hooks/utils/cache/cacheDefaults.ts]: Talked on standup. Pending David's incoming caching re-work. I've reverted this commit
@decentdao/engineering

**Comment by Da-Colon** [comment]
@adamgall were you going to dismiss your review?

**Comment by adamgall** [comment]
@Da-Colon i haven't left a review on this PR yet

**Comment by Da-Colon** [comment]
> @Da-Colon i haven't left a review on this PR yet

O sorry, meant the 'changes requested`.

**Comment by tomstuart123** [comment]
love this @DarksightKellar and thanks for incorporating the 'loading' piece. 

I've tested pretty substantially with Shutter, Myosin, Decent DAO home pages. It makes a huge difference and was much needed for the experience.

Approved

**Comment by adamgall** [comment]
@Da-Colon sorry, I didn't realize i had a stale review on this PR. My fault for any holdups that caused.

**Pull request merged**